src/install/lockfile/Tree.zig

```diff
@@ -534,9 +551,48 @@ pub fn processSubtree(
             }
         }
 
-        const dependency = builder.dependencies[dep_id];
-
         const hoisted: HoistDependencyResult = hoisted: {
+            if (comptime method == .filter) {
+                // not filtered, but does it match a nohoist pattern?
+                if (builder.manager.nohoist_patterns.len != 0) try_nohoist: {
+                    const string_buf = builder.lockfile.buffers.string_bytes.items;
+
+                    try dep_subpath.ensureTotalCapacity(subpath.items().len + @intFromBool(subpath.items().len != 0) + dependency.name.len());
+                    dep_subpath.appendSliceAssumeCapacity(subpath.items());
+                    if (subpath.items().len != 0) {
+                        dep_subpath.appendAssumeCapacity('/');
+                    }
+                    dep_subpath.appendSliceAssumeCapacity(dependency.name.slice(string_buf));
+
+                    if (dependency.version.tag == .workspace) {
+                        break :try_nohoist;
+                    }
+
+                    for (builder.manager.nohoist_patterns) |nohoist_pattern| {
+                        if (glob.match(nohoist_pattern, dep_subpath.items()).matches()) {
+                            // make sure it's not circular
+                            var curr = trees[next.id].parent;
+                            while (curr != invalid_id) {
+                                const curr_dep_id = trees[curr].dependency_id;
+                                const curr_pkg_id = switch (curr_dep_id) {
+                                    root_dep_id => 0,
+                                    else => builder.resolutions[curr_dep_id],
+                                };
+                                var curr_resolutions = builder.resolution_lists[curr_pkg_id];
+                                for (curr_resolutions.begin()..curr_resolutions.end()) |tree_dep_id| {
+                                    const res_id = builder.resolutions[tree_dep_id];
+                                    if (res_id == pkg_id) {
+                                        break :try_nohoist;
```

Comment:
Are you sure this is the right approach? I could be misunderstanding the code, but it seems that, in-case of a circular dependency, we silently fall back to hoisting.

Is this what the user wants? If I were a user and I specifically told my build system to **not hoist my damn dependency**, I wouldn't want there to ever be a case in which the package manager decided to hoist it, especially silently.

I would suggest that we throw an error saying:

> Hey you wanted `nohoist` on this but because there's a circular dep, we can't really do that. We suggest you remove `nohoist` and try again.